### PR TITLE
Some fixes

### LINKED ISSUE #1: Feedback
Hi @myk0ls, thank you for providing your solution, it is definitely on the right track.
We have some feedback for you and would like you to take the time and try to implement some of these changes.

---

What we liked:

1. Kudos for writing it in Scala.
2. Nice use of external libraries.
3. Good usage of case classes.
4. Validation of some data types like lat, lng.
5. Proper `.gitignore`.

---

General issues:

1. I believe the `GeoPoint` constructor (`apply()` method) never actually gets used. Besides, it has `Type missmatch` errors, but the compiler seems to ignore it. Is there a reason why `GeoPoint` isn't just a normal case class?
2. It's nice that you implemented your own ray casting algorithm, but we believe it has issues in the logic. Try to cover all possible edge cases of the algorithm with unit tests, you can even try to talk to an AI about the potential issues that you may have in it. Also you have some comments there, but for something rather complicated like this, I'd expect a more detailed description of the logic inside the comments.
3. A lot of commented code.
4. Are all `final` qualifiers necessary?
5. No unit tests yet.
6. The formatting is a bit off, I would suggest using something like `scalafmt`, though it's not a dealbreaker.
7. I'm not a fan of param names like this `Edge(_1: GeoPoint, _2: GeoPoint)` - `_1` is a bit hard to read. You might have done it this way because that is how scala handles tuples, but even then, we tend to deconstruct the tuples into properly named values.
8. Nested `List[List[]]` types. Not a huge deal, it's only a 2D list, but if you could write code in a way where you didn't have nested lists, that would be a very big bonus.
9. You could use a CLI args parsing library, instead of writing your own.
10. Suggestion - use custom types in places like this, instead of just strings, cause otherwise it's very easy to mix them up when calling the function `def writeJson(filePath: String, data: String)`.
11. Incorrect package names in at least one place.
12. No error handling in some places (`IOProcessor` `os.read` for example).
13. `val write` in `Main` is just not necessary, if it's a simple `Unit`.

---

It would be best if you provided your changes in a pull request, that way we leave comments.

If you have any questions, feel free to reply to this issue.

### PR DESCRIPTION
**Pull request by myk0ls**
Unit tests, CLI argument parsing library, some error handling.

Things I still haven't done: rewriting nested lists, custom types (made few of them, need to implement them).

In terms of the actual logic/algorithm I did a few changes. Instead of connecting polygon's end automatically, I made it so it gets filtered out as an invalid polygon.  In ray-casting algorithm I renamed the variables, added an edge case for a horizontal edge and AI recommended to manage the edge case of a point being aligned with a vertex, so that only the edge's lower point is considered when doing checks.

### CODE REVIEW

File: src/main/scala/logic/GeoProcessor.scala [hunk 1]
```diff
@@ -23,14 +19,20 @@ object GeoProcessor {
...
 
     results
-  } 
-  
+  }
+
   def toPolygons(region: Region): List[Polygon] = {
     region.coordinates.map(polygon => toPolygon(region.name, polygon))
+      .filter(polygon => polygon.edges.size >= 3)
+      .filter(isClosed)
```
Comment (arnasbr) [inline src/main/scala/logic/GeoProcessor.scala]: filtering twice is not very efficient

File: src/main/scala/models/Region.scala [hunk 2]
```diff
@@ -3,6 +3,6 @@ import upickle.default._
 import spatial.GeoPoint
 
 case class Region(name: String, coordinates: List[List[GeoPoint]])
```
Comment (arnasbr) [inline src/main/scala/models/Region.scala]: as you mentioned, you still use nested lists here, which is not ideal

File: src/main/scala/spatial/GeoPoint.scala [hunk 3]
```diff
@@ -2,38 +2,35 @@ package spatial
...
+  )
+
+  def create(longitude: Double, latitude: Double): Option[GeoPoint] = {
+    if (
+      validateLongitude(longitude) &&
+      validateLatitude(latitude)
+    ) Some(GeoPoint(longitude, latitude))
+    else
+      None
+  }
```
Comment (arnasbr) [inline src/main/scala/spatial/GeoPoint.scala:26]: Seems like this method is never used

File: src/main/scala/spatial/GeoPoint.scala [hunk 4]
```diff
@@ -2,38 +2,35 @@ package spatial
 import upickle.default._
 import models._
 
-final case class GeoPoint (coordinates: (Double, Double))
+case class GeoPoint(coordinates: (Double, Double))
```
Comment (Aleksandr1802) [inline src/main/scala/spatial/GeoPoint.scala]: Is it necessary to have coordinates as `tuple`? Maybe `lat`, `lon` can be more readable?

File: src/main/scala/logic/IOProcessor.scala [hunk 5]
```diff
@@ -3,31 +3,54 @@ package logic
...
-    upickle.default.read[List[A]](ujson.read(jsonStr))
+  def decodeJson[A: ReadWriter](jsonStr: String): Try[List[A]] = {
+    val result = Try(upickle.default.read[List[A]](ujson.read(jsonStr)))
+
+    result match {
+      case Success(value) =>
+        Success(value)
+      case Failure(exception) =>
+        Failure(exception)
+    }
```
Comment (arnasbr) [inline src/main/scala/logic/IOProcessor.scala]: what is the point of this :upside_down_face:

Comment (arnasbr) [reply src/main/scala/logic/IOProcessor.scala]: same for other similar matches

File: src/main/scala/logic/PathFile.scala [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package logic
+
+import scala.util.Try
+import os._
+import scala.util.Success
+import scala.util.Failure
+
+case class PathFile(path: String)
```
Comment (arnasbr) [inline src/main/scala/logic/PathFile.scala]: `PathFile` is an okay approach, but you can just use already existing data types, like java / scala `Path` for example

File: src/test/scala/IOTests.scala [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  test("writeJson - correct path") {
+    val tempDir: Path  = os.temp.dir(testPath)
+    val testFile: Path = os.temp("", tempDir)
+
+    val result = IOProcessor.writeJson(testFile.toString(), "This is a test!")
+
+    assert(result.isSuccess)
+
+    os.remove.all(tempDir)
+  }
```
Comment (arnasbr) [inline src/test/scala/IOTests.scala]: are you sure os.remove.all will be called if assertion fails? I haven't used `munit` myself so I'm not sure, but I have a suspicion this might be a minor issue

File: src/main/scala/logic/PathFile.scala (hunk 6 above)
Comment (Aleksandr1802) [inline src/main/scala/logic/PathFile.scala]: What's the purpose/idea of that class?

Comment (arnasbr) [reply src/main/scala/logic/PathFile.scala]: In fact, it seems PathFile is not used anywhere

File: src/main/scala/logic/IOProcessor.scala [hunk 8]
```diff
@@ -3,31 +3,54 @@ package logic
...
-  //   if (os.exists(path))
-  //     Some(os.read(path))
-  //   else
-  //     None
-  // }
-
-  def readJson(filePath: String): String = {
-    val path = Path(filePath, os.pwd)
-    os.read(path)
+  def readJson(filePath: String): Try[String] = {
```
Comment (arnasbr) [inline src/main/scala/logic/IOProcessor.scala]: for file paths, use existing data types like Path from scala / java

**Comment by arnasbr** [comment]
Generally quite nice changes - algorithm improved, tests added, good job! But we would like to see the other mentioned points fixed as well, especially the nested lists one

**Comment by myk0ls** [comment]
Tried my hand at fixing the issues.
Also added more error handling in main. Sorry in advance for a lot of nested match cases.
Furthermore, when I was rewriting deserialization of GeoPoint to use custom .create() method, I just put the .getOrElse for None handling reverting to a default GeoPoint value (obviously not great, I'll still need to add a way to maybe dismiss locations that might have invalid data).
For the nested lists just created a new separate case class GeoArea, but I was wondering if skipping this bridge class and decoding straight into a Polygon, building edges would be more logical?
